### DISCUSSION TITLE
Usage of native/viper emitters for libraries

### DISCUSSION THREAD

**Discussion by ThinkTransit** [micropython/micropython#11157] [General]
If I wanted to try and optimise a library, would I need modify and decorate every function or is there a shortcut?

I did a simple test and all three returned the same time so I imagine the emitter only applies to the code immediately inside the decorated function?

```python
import datetime
from utils import timed_function
from nsw_tz import AESTTimeZone

TZ = AESTTimeZone()

@timed_function
def foo():
    time.strftime('%Y-%m-%d %H:%M:%S', datetime.datetime.now(tz=TZ).tuple())
    
@timed_function
@micropython.native
def foo_native():
    time.strftime('%Y-%m-%d %H:%M:%S', datetime.datetime.now(tz=TZ).tuple())
    
@timed_function
@micropython.viper
def foo_viper():
    time.strftime('%Y-%m-%d %H:%M:%S', datetime.datetime.now(tz=TZ).tuple())
    

>>> foo()
Function foo Time = 176.304ms

>>> foo_native()
Function  Time = 176.459ms

>>> foo_viper()
Function  Time = 176.630ms

>>> 
```

**Comment by rkompass** [comment]
Your assumption is correct.

**Comment by Josverl** [comment]
Optimize for speed, or size? , you get to pick one only.

Assuming you are after speed, there are more options than just adding a decorator.

https://docs.micropython.org/en/latest/reference/speed_python.html#maximising-micropython-speed

**Reply by ThinkTransit** [reply]
Mainly speed.

Thanks yes I have been working through the additional items there.

I was intrigued by doubling in speed.
```
The trade-off for the improved performance (roughly twice as fast as bytecode) is an increase in compiled code size.
```

Considering the native emitter complies to standard python, why not apply it to everything?

**Reply by peterhinch** [reply]
> Considering the native emitter complies to standard python, why not apply it to everything?

The native emitter produces larger code than bytecode.

In my experience the native emitter often does not give much of a speed boost. Viper can be astonishingly fast if applied to the right type of code, notably bit manipulations on integers where it can be almost as fast as assembler.

Like all optimisations the emitters are best applied selectively. An application will usually have a single bottleneck. First identify where it is, measure performance, and apply optimisations measuring at each stage. See also [opt_level](http://docs.micropython.org/en/latest/library/micropython.html?highlight=opt_level#micropython.opt_level).

**Reply by ThinkTransit** [reply]
Thanks for the tips @peterhinch
